### PR TITLE
#8460: Enable Clang-17

### PR DESCRIPTION
**Pull request by vtangTT**
~~blocked by: https://github.com/tenstorrent/tt-metal/issues/8579~~

running final pipelines...

Pipelines w Clang-17:
- post-commit: https://github.com/tenstorrent/tt-metal/actions/runs/9229314005
- nightly fd: https://github.com/tenstorrent/tt-metal/actions/runs/9199329749
- device-perf: https://github.com/tenstorrent/tt-metal/actions/runs/9203282435
- models-perf: https://github.com/tenstorrent/tt-metal/actions/runs/9203284969
- microbenchmarks: https://github.com/tenstorrent/tt-metal/actions/runs/9203283633
- TG frequent: https://github.com/tenstorrent/tt-metal/actions/runs/9201943927
- T3K frequent: https://github.com/tenstorrent/tt-metal/actions/runs/9201947767

Post-commit w GCC9 (current compiler) + these changes: https://github.com/tenstorrent/tt-metal/actions/runs/9211901494

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,8 @@
 cmake_minimum_required(VERSION 3.16)
 cmake_policy(VERSION 3.16)
 
+set(CMAKE_CXX_COMPILER "/usr/bin/clang++-17")
```
Comment (TT-billteng) [inline CMakeLists.txt:4]: pls fix

File: tests/tt_metal/test_utils/packing.hpp [hunk 2]
```diff
@@ -43,7 +43,7 @@ std::vector<PackType> pack_vector(const std::vector<ValueType>& values) {
     unsigned int index = 0;
     std::for_each(results.begin(), results.end(), [&](PackType& result) {
         for (unsigned j = 0; j < num_values_to_pack; j++) {
-            result |= values[index].to_packed() << (index * ValueType::SIZEOF * CHAR_BIT);
+            result |= values[index].to_packed() << (j * ValueType::SIZEOF * CHAR_BIT);
```
Comment (TT-billteng) [inline tests/tt_metal/test_utils/packing.hpp:46]: [@kkwong10 ](https://github.com/kming) tsk tsk

File: CMakeLists.txt (hunk 1 above)
Comment (vtangTT) [reply CMakeLists.txt]: fixed

**Comment by TT-billteng** [comment]
do these changes still work with default Ubuntu gcc?

**Comment by vtangTT** [comment]
> do these changes still work with default Ubuntu gcc?

It will build, but I haven't run post-commit with gcc-9 + these changes

**Comment by TT-billteng** [comment]
> > do these changes still work with default Ubuntu gcc?
> 
> It will build, but I haven't run post-commit with gcc-9 + these changes

Can we try 1 run to get an idea?

**Review by TT-billteng — APPROVED** [review]

**Comment by vtangTT** [comment]
Pipelines w Clang-17:
- post-commit: https://github.com/tenstorrent/tt-metal/actions/runs/9200186861
- nightly fd: https://github.com/tenstorrent/tt-metal/actions/runs/9199329749
- device-perf: https://github.com/tenstorrent/tt-metal/actions/runs/9203282435
- models-perf: https://github.com/tenstorrent/tt-metal/actions/runs/9203284969
- microbenchmarks: https://github.com/tenstorrent/tt-metal/actions/runs/9203283633
- TG frequent: https://github.com/tenstorrent/tt-metal/actions/runs/9201943927
- T3K frequent: https://github.com/tenstorrent/tt-metal/actions/runs/9201947767

Post-commit w GCC9 (current compiler) + these changes: https://github.com/tenstorrent/tt-metal/actions/runs/9211901494

File: .github/workflows/build-artifact.yaml [hunk 3]
```diff
@@ -33,7 +33,7 @@ jobs:
       - uses: ./.github/actions/install-python-deps
       - name: Build tt-metal and libs
         run: |
-          cmake -B build -G Ninja
+          cmake -B build -G Ninja -DCMAKE_CXX_COMPILER=clang++-17
```
Comment (tt-rkim) [inline .github/workflows/build-artifact.yaml:36]: Any plans to make this default? @TT-billteng

File: .github/workflows/build.yaml [hunk 4]
```diff
@@ -79,6 +80,6 @@ jobs:
         run: tar -xvf ttm_${{ matrix.arch }}-${{ matrix.config}}.tar
       - name: Build tt-metal tests
         run: |
-          rm -f build/CMakeCache.txt
-          cmake -B build -G Ninja
-          cmake --build build --target tests -- -j`nproc`
+          rm -f build/CMakeCache.txt && find build -name 'cmake_pch.hxx.pch' -exec touch {} \;
```
Comment (tt-rkim) [inline .github/workflows/build.yaml:83]: Is this to force PCH recompile? Why is this needed if they don't change between the lib and test build?

File: setup.py [hunk 5]
```diff
@@ -107,6 +107,7 @@ def get_build_env():
             **os.environ.copy(),
             "TT_METAL_HOME": Path(__file__).parent,
             "TT_METAL_ENV": "production",
+            "CXX": "clang++-17",
```
Comment (tt-rkim) [inline setup.py:110]: This passes it in ? Doesn't need as a `-D` arg?

File: tests/tt_metal/test_utils/packing.hpp (hunk 2 above)
Comment (tt-rkim) [reply tests/tt_metal/test_utils/packing.hpp:46]: DAI LO NO CODE

File: ttnn/cpp/ttnn/operations/data_movement.hpp [hunk 6]
```diff
@@ -81,7 +81,7 @@ inline ttnn::Tensor permute(
         std::vector<uint32_t> full_shape_vec{};
         int i = 0;
         while(i < 3 and shape[i] == 1) i++;
-        for(i; i < shape.rank(); i++) {
+        for(; i < shape.rank(); i++) {
```
Comment (tt-rkim) [inline ttnn/cpp/ttnn/operations/data_movement.hpp:84]: LMfao wut

File: setup.py (hunk 5 above)
Comment (vtangTT) [reply setup.py:110]: yep

File: .github/workflows/build.yaml (hunk 4 above)
Comment (vtangTT) [reply .github/workflows/build.yaml:83]: It's to tell clang that we don't need to recompile PCH, otherwise it'll complain it's outdated even though nothing changed :(

**Review by tt-rkim — APPROVED** [review]

File: .github/workflows/build.yaml [hunk 7]
```diff
@@ -79,6 +80,6 @@ jobs:
...
+          cmake -B build -G Ninja -DCMAKE_CXX_COMPILER=clang++-17
```
Comment (ayerofieiev-tt) [inline .github/workflows/build.yaml:84]: I think it is better to specify cxx version in the cmake file itself

# Set the C++ compiler
```
set(CMAKE_CXX_COMPILER "clang++-17")

# Other project settings
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```

or maybe even start using a toolchain file, preparing to package management onboarding
`CMAKE_TOOLCHAIN_FILE`

File: .github/workflows/build-artifact.yaml (hunk 3 above)
Comment (TT-billteng) [reply .github/workflows/build-artifact.yaml:36]: I don't want to force the user to be using clang++17. If/when we move to ubuntu 24.04, gcc 13 may give better perf numbers. So at best I'd like to "recommend" the compiler of choice as of now

**Review by arakhmati — APPROVED** [review]

**Review by tt-asaigal — APPROVED** [review]

File: CMakeLists.txt [hunk 8]
```diff
@@ -1,13 +1,29 @@
 cmake_minimum_required(VERSION 3.16)
 cmake_policy(VERSION 3.16)
 
+############################################
```
Comment (ayerofieiev-tt) [inline CMakeLists.txt:4]: remove?

**Review by ntarafdar — APPROVED** [review]

File: .github/workflows/build.yaml (hunk 7 above)
Comment (vtangTT) [reply .github/workflows/build.yaml:84]: If we set it, we won't be able to override clang17 through command line

I agree we should look into a toolchain file in the future.
